plugin/restore/mysql/mysql.go

```diff
@@ -528,68 +539,113 @@ func (r *Restore) FetchArchivedBinlogFiles(ctx context.Context) error {
+			binlogFilesOnServerMap[file.Name] = file
 		}
-
-		localBinlogSize, ok := localFileMap[serverBinlog.Name]
-		if !ok {
-			todo[serverBinlog.Name] = true
-			continue
+		for _, file := range binlogFilesLocalSorted {
+			fileOnServer, existOnServer := binlogFilesOnServerMap[file.Name]
+			if existOnServer && file.Size != fileOnServer.Size {
+				path := filepath.Join(r.binlogDir, file.Name)
+				log.Debug("Deleting inconsistent local binlog file",
+					zap.String("path", path),
+					zap.Int64("sizeLocal", file.Size),
+					zap.Int64("sizeOnServer", fileOnServer.Size))
+				if err := os.Remove(path); err != nil {
+					log.Error("Failed to remove inconsistent local binlog file", zap.String("path", path), zap.Error(err))
+					return fmt.Errorf("failed to remove inconsistent local binlog file[%s], error[%w]", path, err)
+				}
+				if err := r.downloadBinlogFile(ctx, fileOnServer, file.Name == latestBinlogFileOnServer.Name); err != nil {
+					log.Error("Failed to re-download inconsistent local binlog file", zap.String("path", path), zap.Error(err))
+					return fmt.Errorf("failed to re-download inconsistent local binlog file[%s], error[%w]", path, err)
+				}
+			}
 		}
 
-		if localBinlogSize != serverBinlog.Size {
-			// exist on local and file size not match, delete and then download it
-			log.Warn("inconsistent binlog file size detected", zap.String("binlogFile", serverBinlog.Name), zap.Int64("server", serverBinlog.Size), zap.Int64("local", localBinlogSize))
-			if err := os.Remove(filepath.Join(r.binlogDir, serverBinlog.Name)); err != nil {
-				return fmt.Errorf("cannot remove %s, error: %w", serverBinlog.Name, err)
+		// Check whether we can only use local binlog files to restore to targetTs:
+		// 1. The local binlog files have no gaps.
+		// 2. Some file's first binlog event ts is already larger than targetTs.
+		// 3. The first local binlog file <= first server binlog file
+		// If all true, we can skip downloading binlog files at all.
+		log.Debug("Checking whether we can just replay existing local binlog files to restore")
+		if binlogFilesLocalSorted[0].Seq <= binlogFilesOnServerSorted[0].Seq {
+			isContinuous := binlogFilesAreContinuous(binlogFilesLocalSorted)
+			containsTargetTs := false
+			if isContinuous {
+				for _, file := range binlogFilesLocalSorted {
+					eventTs, err := r.parseFirstLocalBinlogEventTimestamp(ctx, file.Name)
+					path := filepath.Join(r.binlogDir, file.Name)
+					if err != nil {
+						log.Error("Failed to parse the first binlog event timestamp for local binlog file",
+							zap.String("path", path),
+							zap.Error(err))
+						return fmt.Errorf("failed to parse the first binlog event timestamp for local binlog file[%s], error[%w]", path, err)
+					}
+					if eventTs > targetTs {
+						log.Debug("Found a local binlog file whose first event's timestamp is larger than targetTs",
+							zap.String("path", path),
+							zap.Int64("targetTs", targetTs),
+							zap.Int64("eventTs", eventTs))
```

Comment:
There is no need to break, we can just return here

Reply:
Done.